### PR TITLE
feat(electrum)!: Update `bdk_electrum` to use merkle proofs

### LINKED ISSUE #980: Update `bdk_electrum` to use electrum's merkle proof API
**Describe the enhancement**

As mentioned by @LLFourn in https://github.com/bitcoindevkit/bdk/pull/976#discussion_r1193920143:

> ... I think electrum really intends for you to use the merkle proof API. Best solution is to do that and attach confirmation time of each block to `LocalChain`.

**Additional context**

* This should be attempted after #976 is merged.
~* This may require changing `LocalChain`.~

### PR DESCRIPTION
**Pull request by LagginTimes**
Fixes #980.

### Description

This PR is the first step in reworking `bdk_electrum` to use merkle proofs. When we fetch a transaction, we now also obtain the merkle proof and block header for verification. We then insert an anchor only after validation that the transaction exists in a confirmed block. The loop logic that previously existed in `full_scan` to account for re-orgs has also been removed as part of this rework.

This is a breaking change because `graph_update`s now provide the full `ConfirmationTimeHeightAnchor` type. This removes the need for the `ElectrumFullScanResult` and `ElectrumSyncResult` structs that existed only to provide the option for converting the anchor type from `ConfirmationHeightAnchor` into `ConfirmationTimeHeightAnchor`.

### Changelog notice

* `ConfirmationTimeHeightAnchor` and `ConfirmationHeightAnchor` have been removed.
* `ConfirmationBlockTime` has been introduced as a new anchor type.
* `bdk_electrum`'s `full_scan` and `sync` now return `graph_update`s with `ConfirmationBlockTime`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

**Comment by notmandatory** [comment]
Since this doesn't touch the wallet APIs I moved it to the beta milestone.

**Comment by LagginTimes** [comment]
> Since this doesn't touch the wallet APIs I moved it to the beta milestone.

I'm okay with this!

File: crates/electrum/src/bdk_electrum_client.rs
Comment (LLFourn) [inline crates/electrum/src/bdk_electrum_client.rs]: I think we should have a `block_header_cache` here too. Then you can get the last block header we had at that height. If the merkle proof validation fails we can fetch it again, and validate against the new one. This will reduce the number of requests we make to electrum by a bit.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (LLFourn) [inline crates/electrum/src/bdk_electrum_client.rs]: This seems like this is a suboptimal anchor to use since you're entering the same `block_height` twice. It would actually be good if you could delete `ConfirmationTimeHeightAnchor` and `ConfirmationHeightAnchor` from the codebase now since both electrum and esplora will report the exact block that the tx is in with this PR. It only needs to be recorded once.

For now this would be good to use for both:

```rust
struct ConfirmationBlockTime {
    block_id: BlockId,
    confirmation_time: u64
}
```

Later we can add the block position to this and maybe even the proof. 

Deleting `ConfirmationTimeHeight` anchor would need ACK from @evanlinjin  and @notmandatory since it's used by `Wallet`. Just for background the only reason `ConfirmationTimeHeightAnchor` exists is because of how wacky our electrum client was before this PR. I think simplifying the anchors to make use of this should be done ASAP.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (LLFourn) [inline crates/electrum/src/bdk_electrum_client.rs]: Is it possible to use rust bitcoin's merkle proof validator for this? Why does electrum have its own validation logic?

**Review by LLFourn — COMMENTED** [review]
Thanks this simplifies a lot. Left a few comments around the anchors/merkle proof bit.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (LLFourn) [reply crates/electrum/src/bdk_electrum_client.rs]: If we don't want to delete the old anchors we should still return the correct thing from here. It can be `Into` converted into the old one later.

Comment (evanlinjin) [reply crates/electrum/src/bdk_electrum_client.rs]: ACK

**Comment by notmandatory** [comment]
@LagginTimes oops, I didn't read all the comments before, looks like the wallet API will need to be updated so I've moved this back to the alpha milestone.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (notmandatory) [reply crates/electrum/src/bdk_electrum_client.rs]: I support simplifying the anchors now. I'm happy to do whatever changes are required for to the sqlite storage crate. 

cc: @matthiasdebernardini this anchor change will also impact the postgres store you're working on.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (LagginTimes) [reply crates/electrum/src/bdk_electrum_client.rs]: The validation logic being used right now was implemented as a utility in `rust-electrum-client`, which was needed for LDK's electrum integration:
https://github.com/bitcoindevkit/rust-electrum-client/pull/122

Would using rust bitcoin be more advantageous in this scenario?

Comment (LLFourn) [reply crates/electrum/src/bdk_electrum_client.rs]: I have no idea. I am just curious why @tnull added that when there is presumably this algorithm is implemented in `rust-bitcoin`.

Comment (tnull) [reply crates/electrum/src/bdk_electrum_client.rs]: IIRC, Electrum's `get_merkle`/`GetMerkleRes` doesn't provide sufficient/slightly different data, so that we couldn't construct `bitcoin::MerkleBlock`/ `PartialMerkleTree` to use upstream's validation logic (which of course would have been our preferred way).

File: crates/chain/src/chain_data.rs
Comment (ValuedMammal) [inline crates/chain/src/chain_data.rs:157]: docs: this line no longer applies

> Note that the confirmation block and the anchor block can be different here.

File: crates/chain/tests/test_tx_graph.rs
Comment (ValuedMammal) [inline crates/chain/tests/test_tx_graph.rs]:
```suggestion
                block_id: tip.get(ht).unwrap().block_id(),
```

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs]:
```suggestion
        for (keychain, spks) in request.spks_by_keychain {
```

Comment (ValuedMammal) [reply crates/electrum/src/bdk_electrum_client.rs]: 📌  The docs for `full_scan` are a bit lacking with respect to `fetch_prev_txouts`, see https://github.com/bitcoindevkit/bdk/pull/1494#issuecomment-2204199786

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs]: docs: "cps" and "tx_cache" are no longer relevant parameters, same for `populate_with_outpoints` below

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs:373]: question: Is there a scenario where the `validate_merkle_for_anchor` function could fail and we'd want to know about it? in other words should we be worried if validating the proof fails even after updating the header?

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs]: if we just use `.iter()` we can remove the call to `clone`

File: crates/electrum/tests/test_electrum.rs
Comment (ValuedMammal) [inline crates/electrum/tests/test_electrum.rs:378]: We should be able to make assertions on the expected graph anchors, something like:

```diff
     // Create transactions that are tracked by our receiver.
+    let mut txids = vec![];
+    let mut blockhashes = vec![];
     for _ in 0..REORG_COUNT {
-        env.send(&addr_to_track, SEND_AMOUNT)?;
-        env.mine_blocks(1, None)?;
+        txids.push(env.send(&addr_to_track, SEND_AMOUNT)?);
+        blockhashes.extend(env.mine_blocks(1, None)?);
     }
 
     // Sync up to tip.
     env.wait_until_electrum_sees_block()?;
     let update = client.sync(
@@ -163,10 +161,17 @@ fn tx_can_become_unconfirmed_after_reorg() -> anyhow::Result<()> {
         .map_err(|err| anyhow::anyhow!("LocalChain update error: {:?}", err))?;
     let _ = recv_graph.apply_update(update.graph_update.clone());
 
     // Retain a snapshot of all anchors before reorg process.
     let initial_anchors = update.graph_update.all_anchors();
+    let anchors: Vec<_> = initial_anchors.iter().cloned().collect();
+    assert_eq!(anchors.len(), REORG_COUNT);
+    for i in 0..REORG_COUNT {
+        let (anchor, txid) = anchors[i];
+        assert_eq!(anchor.anchor_block().hash, blockhashes[i]);
+        assert_eq!(txid, txids[i]);
+    }
```

**Review by ValuedMammal — COMMENTED** [review]
I did a first pass, mostly documentation nits. Sorry I was reviewing while you were pushing, so comments might be outdated

**Comment by notmandatory** [comment]
@LagginTimes hey sorry I just merged some smaller PRs and now this one needs a rebase.

**Comment by evanlinjin** [comment]
I think we should not wait for complete tests before merging this PR. As long as the examples work fine it is good. This is for "beta" which does not need to be bug free. Also having this PR merged means we can work on the other anchor changes (which I need to create a ticket for).

**Comment by evanlinjin** [comment]
Also, the anchor changes that I am suggesting is going to conflict with `ConfirmationBlockTime`.

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs:227]: 💯

File: crates/electrum/src/bdk_electrum_client.rs
Comment (ValuedMammal) [inline crates/electrum/src/bdk_electrum_client.rs:454]: nit, feel free to ignore

```suggestion
        .filter(|(&height, _)| Some(height) > agreement_height)
```

File: crates/electrum/tests/test_electrum.rs
Comment (ValuedMammal) [inline crates/electrum/tests/test_electrum.rs:25]: 📌 If I recall, the name of this test is a relic from before the `sync` and `full_scan` concepts were developed. I might now call it `sync_should_update_chain_and_tx_graph` or similar, but not terribly important at the moment.

**Review by ValuedMammal — APPROVED** [review]
ACK 1a62488abfd6213c2780c1465aebe01dfbb6b20a

File: crates/electrum/src/bdk_electrum_client.rs
Comment (notmandatory) [reply crates/electrum/src/bdk_electrum_client.rs:373]: 📌  This is a good question and it makes sense to have a runtime option or feature to throw an error if merkle proof validation fails. But I don't think this has to be done in this PR or for the beta release since it will only change the API for the `bdk_electrum` crate. @ValuedMammal would you mind just making an issue for this?

**Review by notmandatory — APPROVED** [review]
ACK 1a62488abfd6213c2780c1465aebe01dfbb6b20a

Thanks for also adding the extra esplora tests for electrum!

**Pull request merged**
